src/metabase/driver/sql_jdbc/actions.clj

```diff
@@ -23,41 +24,72 @@
+   (java.sql Connection PreparedStatement SQLException)))
 
 (set! *warn-on-reflection* true)
 
-(defmulti parse-sql-error
-  "Parses the raw error message returned after an error in the driver database occurs, and converts it into a sequence
-  of maps with a :column and :message key indicating what went wrong."
-  {:arglists '([driver database e]), :added "0.44.0"}
-  driver/dispatch-on-initialized-driver
+;;; +----------------------------------------------------------------------------------------------------------------+
+;;; |                                               Error handling                                                   |
+;;; +----------------------------------------------------------------------------------------------------------------+
+
+(defmulti maybe-parse-sql-error
+  "Try to parse the error message returned by JDBC driver.
+
+  The methods should returns a map of:
+  - type: the error type. Check [[metabase.actions.error]] for the full list
+  - message: a nice message summarized of what went wrong
+  - errors: a map from field-name => sepcific error message. This is used by UI to display per fields error."
+  {:arglists '([driver error-type database error-message]), :added "0.48.0"}
+  (fn [driver error-type _database _error-message]
+   [(driver/dispatch-on-initialized-driver driver) error-type])
   :hierarchy #'driver/hierarchy)
 
-(defn- parse-error
-  "Returns errors in a way that indicates which column had the problem. Can be used to highlight errors in forms."
+(defmethod maybe-parse-sql-error :default
+  [_driver _error-type _database _e]
+  nil)
+
+(defn- driver->available-error-parser
+  [driver]
+  (->> (keys (methods maybe-parse-sql-error))
+       (filter (fn [x] (and (sequential? x)
+                            (isa? driver/hierarchy driver (first x)))))
+       (map second)))
+
+(defn- parse-sql-error
   [driver database e]
-  (let [default-method? (= (get-method parse-sql-error driver)
-                           (get-method parse-sql-error :default))]
-    (if default-method?
-      (ex-data e)
-      (let [message (ex-message e)]
-        (if-let [parsed-errors (when message
-                                 (try
-                                   (parse-sql-error driver database message)
-                                   (catch Throwable e
-                                     (log/error e (trs "Error parsing SQL error message {0}: {1}" (pr-str message) (ex-message e)))
-                                     nil)))]
-          {:errors (into {} (map (juxt :column :message)) parsed-errors)}
-          {:message (or message (pr-str e))})))))
-
-(defn- catch-throw [e status-code & [more-info]]
-  (throw
-   (ex-info (ex-message e)
-            (merge {:exception-data (ex-data e)
-                    :status-code status-code}
-                   more-info)
-            e)))
+  (try
+   (some #(maybe-parse-sql-error driver % database (ex-message e)) (driver->available-error-parser driver))
```

Comment:
This could be simplified. `(methods maybe-parse-sql-error)` returns a dispatch table of driver and error-type to parser function. You're effectively looking up this table twice just to get the parser functions. Once to get the error-types for a driver, and then using those error-types, you get the methods. But you could just do it in one step, where you get the methods for a given driver and apply them directly.

e.g.

```clojure
(defn- parse-sql-error
  [driver database e]
  (let [parsers-for-driver (keep (fn [[[method-driver _error-type] method]]
                                   (when (= method-driver driver)
                                     method))
                                 (dissoc (methods maybe-parse-sql-error) :default))]
    (try
      (some #(% driver database (ex-message e)) parsers-for-driver)
      ;; Catch errors in parse-sql-error and log them so more errors in the future don't break the entire action.
      ;; We'll still get the original unparsed error message.
      (catch Throwable new-e
        (log/error new-e (trs "Error parsing SQL error message {0}: {1}" (pr-str (ex-message e)) (ex-message new-e)))
        nil))))
```